File: Arrays/rotateArray.java
/*
  LeetCode 189. Rotate Array: https://leetcode.com/problems/rotate-array/
*/
class Solution {
	/*
	  Time: O(n)
      Space: O(n)
	*/
    public void rotate(int[] nums, int k) {
        int[] temp = new int[nums.length];
        for (int i = 0; i < nums.length; i++) {
            temp[(i + k) % nums.length] = nums[i];
        }
        for (int i = 0; i < nums.length; i ++) {
            nums[i] = temp[i];
        }
    }

    /*
      Space improved solution
      Time: O(n)
      Space: O(1)
    */
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k %= n;
        reverse(nums, 0, n - 1);
        reverse(nums, 0, k - 1);
        reverse(nums, k, n - 1);
    }
    
    public void reverse(int[] nums, int start, int end) {
        while (start < end) {
            int temp = nums[start];
            nums[start++] = nums[end];
            nums[end--] = temp;
        }
    }}

File: Arrays/rotateArray.py
# LeetCode 189. Rotate Array: 189. Rotate Array
# Time: O(n)
# Space: O(1)

class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        k %= n
        self.reverse(nums, 0, n - 1)
        self.reverse(nums, 0, k - 1)
        self.reverse(nums, k, n - 1)
        
    def reverse(self, nums: List[int], start: int, end: int) -> None:
        """
        reverse a array (end inclusively)
        """
        while start < end:
            nums[start], nums[end] = nums[end], nums[start]
            start += 1
            end -= 1